specs-derive/src/impl_saveload.rs

```diff
@@ -0,0 +1,496 @@
+                    .cloned()
+                    .enumerate()
+                    .map(|(i, _)| Some(Ident::new(&format!("field{}", i), data.enum_token.0)))
+                    .collect();
+
+                let field_ids_2 = field_ids.clone();
+
+                big_match_de = quote!{
+                    #big_match_de
+                    #saveload_name::#ident( #( #field_ids ),* ) => #name::#ident( #( ::specs::saveload::FromDeserialize::from(#field_ids_2, &mut ids)? ),* ),
+                };
+            }
+            Fields::Unit => {
+                big_match_ser = quote! {
+                    #big_match_ser
+                    #name::#ident => #saveload_name::#ident,
+                };
+
+                big_match_de = quote! {
+                    #big_match_de
+                    #saveload_name::#ident => #name::#ident,
+                };
+            }
+        }
+    }
+
+    let ser = quote!{
+        Ok(match *self {
+            #big_match_ser
+        })
+    };
+
+    let de = quote!{
+        Ok(match data {
+            #big_match_de
+        })
+    };
+
+    SaveloadDerive {
+        type_def: enum_def,
+        ser,
+        de,
+        saveload_name,
+        saveload_generics: saveload_generics.clone(), // dumb borrow checker, don't feel like fixing because it's probably NBD
+    }
+}
+
+// Edits the bounds of the input type to ensure all fields have `bound`.
+fn add_bound(generics: &mut Generics, bound: TypeParamBound) {
+    use syn::GenericParam;
+    for param in &mut generics.params {
+        if let GenericParam::Type(ref mut type_param) = *param {
+            if type_param.bounds.iter().position(|b| bound == *b).is_some() {
+                continue;
+            }
+            type_param.bounds.push(bound.clone());
+        }
+    }
+}
+
+// Replaces the type with its corresponding `Data` type.
+fn replace_entity_type(ty: &mut Type) {
+    match *ty {
+        Type::Array(ref mut ty) => replace_entity_type(&mut *ty.elem),
+        Type::Tuple(ref mut ty) => {
+            for ty in ty.elems.iter_mut() {
+                replace_entity_type(&mut *ty);
+            }
+        }
+        Type::Paren(ref mut ty) => replace_entity_type(&mut *ty.elem),
+        Type::Path(ref mut ty) => {
+            let ty_tok = ty.clone();
+            *ty = parse_quote!(<#ty_tok as ::specs::saveload::IntoSerialize<MA>>::Data);
+        }
+        Type::Group(ref mut ty) => replace_entity_type(&mut *ty.elem),
+
+        Type::TraitObject(_) => {}
+        Type::ImplTrait(_) => {}
+        Type::Slice(_) => {
+            panic!("Slices are unsupported, use owned types like Vecs or Arrays instead")
+        }
+        Type::Reference(_) => panic!("References are unsupported"),
+        Type::Ptr(_) => panic!("Raw pointer types are unsupported"),
+        Type::BareFn(_) => panic!("Function types are unsupported"),
+        /* We're in a struct so it doesn't matter */
+        Type::Never(_) => unreachable!(),
```

Comment:
When never type is stable this can happen: You could have field with type `!` (Extremely unlikely with struct, more likely with enum. Even more unlikely without generics support).

Reply:
Ah yes, I think I didn't understand the never type. It's meant to be the "official" solution to things like `specs` `NoError` type is now, right? Like a `Result<Foo,!>` is a type that can exist, but its `Err` can never exist at runtime, as guaranteed by its type semantics. And also has the benefit of satisfying every trait because it's a bottom type, so you can use it as shorthand for "don't worry about this case, if we get here we've already panicked anyway."

I got confused because I know `panic!` and its cousins are of type `!` now and assumed it was solely the return type of those devices, I wasn't aware there was a communication to build that into the type system as a feature, neat!

Reply:
Yeah exactly that!